Support constant string concatenation

DeviousStoat opened the pull request:
Implements support for string concatenation in module constants' expressions.

```gleam
const foo = "foo"
const foobar = foo <> "bar"
const foofoobarbaz = foo <> foobar <> "baz"
```

closes #3208

lpil commented on compiler-core/src/erlang.rs:
```diff
@@ -512,6 +512,37 @@ fn tuple<'a>(elems: impl IntoIterator<Item = Document<'a>>) -> Document<'a> {
+
+        Constant::StringConcatenation { .. } => docvec![const_inline(value, env), "/binary"],
+
+        _ => panic!("Constant string concatenation argument can only be a string or const var at code generation time"),
```
Please remove these panics, code generation is not the place to do type checking, and other variants could be suitable in future. Calling `const_inline` on whatever the `Constant::*` is is good.

lpil commented on compiler-core/src/erlang/tests/strings.rs:
```diff
@@ -438,3 +438,32 @@ pub fn main(x) {
 }
+
+#[test]
+fn assert_const_concat() {
```
Could you remove all "foobar" please, it's a rather depressing ww2 reference.

lpil commented on compiler-core/src/erlang/tests/strings.rs:
```diff
@@ -438,3 +438,32 @@ pub fn main(x) {
+        r#"
+const foo = "foo"
+const foobar = foo <> "bar"
+const foofoobarbaz = foo <> foobar <> "baz"
```
Split this into multiple tests please, we only test one example per test.

Could we also have a test for a much larger number of concatenations in one constant too please.

lpil commented on compiler-core/src/format/tests.rs:
```diff
@@ -6156,3 +6156,19 @@ fn newlines_are_not_stripped_if_two_consecutive_anonymous_function_are_passed_as
+  <> " break"
+"#
+    );
+}
```
Can we have tests for unbroken concatenation expressions and also something like `const x = "some long string 1" <> "some long string 2" <> ["here is a list", "with several elements", "in order to make it be too long to fit on one line", "so we can see how it breaks", "onto multiple lines"] <> "and a last string`

lpil commented on compiler-core/src/type_/tests/errors.rs:
```diff
@@ -1905,3 +1905,13 @@ fn leak_multiple_private_types() {
+    assert_module_error!(
+        "
+const foo = 5
+const bar = foo <> \"bar\"
```
Remove the foobar please 🙏

lpil commented on compiler-core/src/javascript/tests/strings.rs:
```diff
@@ -224,3 +224,32 @@ pub fn go(x) {
+fn const_concat() {
+    assert_js!(
+        r#"
+const foo = "foo"
```
Remove the foobar please 🙏

lpil reviewed (commented):
Lovely work! I've left some notes inline. Thank you

DeviousStoat commented:
I think I addressed your comments.
Btw, don't know if it matters too much as I haven't worked on too many big rust projects but I think this codebase is really well made, it's a joy to work on it and easy to pickup, akin to the gleam language itself

lpil commented on compiler-core/src/erlang/tests/snapshots/gleam_core__erlang__tests__strings__assert_const_concat_many_strings.snap:
```diff
@@ -0,0 +1,37 @@
+
+-spec main() -> binary().
+main() ->
+    <<"a"/utf8,
```
This should render like this:

```erlang
main() ->
    <<"a"/utf8,
      "b"/utf8,
      "c"/utf8,
      ...
```

DeviousStoat replied:
ooh right, it's not just format problem, we don't need to `/binary` the concatenation, we can build a bit array of `/utf8` strings. This is a lot nicer now, I am starting to understand what you meant with "string concatenation doesn't require compile time code execution"

lpil commented on compiler-core/src/parse/tests.rs:
```diff
@@ -1121,3 +1121,22 @@ fn newline_tokens() {
+fn const_string_concat_naked_right() {
+    assert_module_error!(
+        "
+const foo = \"foo\" <>
```
Remove the foobar from this file please 🙏

DeviousStoat replied:
oops sorry missed this one

lpil reviewed (commented):
Oh that you, you are very kind! 💜 

This is looking great! Thank you. 2 small thing in the notes

lpil commented on compiler-core/src/erlang.rs:
```diff
@@ -512,6 +512,71 @@ fn tuple<'a>(elems: impl IntoIterator<Item = Document<'a>>) -> Document<'a> {
+        // instead of
+        // <<"a"/utf8,
+        // "b"/utf8>>
+        .set_nesting(INDENT + 2)
```
This will be wrong if the indent level is supposed to be anything other than 1!

For example:

```gleam
const x = [
  <<
    ...,
    ...,
  >>
]
```

I don't mind so much about the exact format used, but the indentation should increase as there's more nesting

DeviousStoat replied:
oh right, I pushed a change with a test case for a long concatenation in a list.
Let me know what you think, it doesn't exactly has the format that you specified before but the indentation should be good now.

I think to render like that:
```erl
main() ->
    <<"a"/utf8,
      "b"/utf8,
      "c"/utf8,
      ...
```
would require changes in the nesting logic or a special case I am not sure, I could look into it if you want

giacomocavalieri replied:
Great job @DeviousStoat!! I think I can help with the formatting here. The code could look something like this (I haven't tested it but this should at least get you closer to a working solution 😁)
```rs
fn const_string_concatenate_bit_array<'a>(
    elems: impl IntoIterator<Item = Document<'a>>,
) -> Document<'a> {
  let elems = elems.collect_vec();
  if elems.is_empty {
    return "<<>>".to_doc();
  }
  
  let open = "<<".to_doc().append(break_("", "").nest(INDENT));
  let close = break_("", "").append(">>".to_doc());
  
  join(elems, break_(", ", ",").nest(INDENT))
    .surround(open, close)
    .group()
}
```

The look will be something like this if the bit array is split:
```erl
<<
  "wibble"/utf8,
  "wobble"/utf8
>>
```

giacomocavalieri commented on compiler-core/src/erlang/tests/snapshots/gleam_core__erlang__tests__strings__assert_const_concat_many_strings_in_list.snap:
```diff
@@ -0,0 +1,37 @@
+
+-spec main() -> list(binary()).
+main() ->
+    [<<"a"/utf8,
+            "b"/utf8,
+            "c"/utf8,
+            "d"/utf8,
+            "e"/utf8,
+            "f"/utf8,
+            "g"/utf8,
+            "h"/utf8,
+            "i"/utf8,
+            "j"/utf8,
+            "k"/utf8,
+            "l"/utf8,
+            "m"/utf8,
+            "n"/utf8,
+            "o"/utf8,
+            "p"/utf8,
+            "q"/utf8,
+            "r"/utf8,
+            "s"/utf8,
+            "t"/utf8,
+            "u"/utf8,
+            "v"/utf8,
+            "w"/utf8,
+            "x"/utf8,
+            "y"/utf8,
+            "z"/utf8>>].
```
I think the preferred look would be:
```erl
<<
    "a"/utf8,
    "b"/utf8,
    ...
>>
```
You can have a look at [this comment](https://github.com/gleam-lang/gleam/pull/3318#discussion_r1664311024) to get an idea how to do it

lpil reviewed (approved):
Bravo!! Thank you
